### ISSUE TITLE
org.osgi.service.component.annotations.Component always require SCR

### ISSUE THREAD

**Issue by laeubi** [osgi/osgi#569]
`org.osgi.service.component.annotations.Component` is annotated with `@RequireServiceComponentRuntime`

This has the implication that one can't use the annotations when the scr component should be optional (e.g. bundle contains code that can be used without DS) because the bundle does not resolve in that case.

I think it would be more appropriate with `resolution = optional` (see also https://github.com/osgi/osgi/discussions/568) so a scr extender is used if available but not required.

**Comment by timothyjward** [comment]
I would be against this change as it is incorrect and hard to use for the majority of bundles, when the purpose of the annotations is to make things easy.

If you have shared code that you want to be able to use in the absence of SCR then separate that code into a different bundle with the relevant package exports. If this really isn’t possible for a specific case then accept that SCR isn’t actually optional as the components are a necessary part of the whole.

**Comment by laeubi** [comment]
> I would be against this change as it is incorrect

Why is is incorrect?

> and hard to use for the majority of bundles

Why is it hard to use? Actually nothing changes here...

> If this really isn’t possible for a specific case then accept that SCR isn’t actually optional as the components are a necessary part of the whole.

I don't think so, see the spec text here:
https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#d0e29610

it does not mandates that a component **must** declare the requirement.

Futher on here in the spec:
https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#service.component-capabilities

> A bundle that contains service components **should require** the osgi.extender capability from SCR. This requirement will wire the bundle to the SCR implementation and ensure that SCR is using the same org.osgi.service.component package as the bundle if the bundle uses that package. 

> The [RequireServiceComponentRuntime](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#org.osgi.service.component.annotations.RequireServiceComponentRuntime) annotation **can be used** to require this capability.

So the spec says it _should_ do so and it _can_ use but it never states it _must_ be used. But making this mandatory when using a `Component` annotations makes them less powerful than the XML ...

So from a programmers POV the `RequireServiceComponentRuntime` is a bit useless if it is mandatory anyways, why should I use it if there is actually no choice?

**Comment by laeubi** [comment]
And just as an example to compare this with a traditional `BundleActivator` used:

If I have a `BundleActivator`, I mostly _want_ my bundle to be started, still it is not _required_ and I can still use classes from that bundle as it resolves, but with the annotation now my bundle never resolves (even if it is never started so no components won't be there anyways) and I can not load any of its classes.

**Comment by maho7791** [comment]
Hi Christoph,

at first, I agree with Tim for the same reasons.


> So the spec says it /should/ do so and it /can/ use but it never 
> states it /must/ be used. But making this mandatory when using a 
> |Component| annotations makes them less powerful than the XML ...
>
https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-components

A component requires the XML for the service description, the Manifest 
header and the implementation. So, if you don't use the annotations and 
still follow these three conditions. You still have a component and no 
mandatory requirement for the SCR extender and your bundle will resolve.

For this case the /should/ in the specification fits.

Using the Component annotations is not mandatory for having DS components.

> So from a programmers POV the |RequireServiceComponentRuntime| is a 
> bit useless if it is mandatory anyways, why should I use it if there 
> is actually no choice?
>
You have another choice, as described above. In addition to that I will 
comment, that you speak "from a programmers POV", which is a 
generalization for all programmers. So, I to understand you right, I 
believe you meant your POV as programmer.

My POV would be different.

If I find code with component annotations, I would expect to have SCR in 
the runtime. The consequence would be, that if I declare Components then 
and SCR would not be resolved, it would confuse me as well.

If I just find a Java class, but the Manifest declaration for components 
and the XML, I would probably think about why no annotations are used? 
But in this case I would know, that I have to take care, that SCR is 
available.

Regards, Mark

> Message ID: ***@***.***> 

-- 
Mark Hoffmann
M.A. Dipl.-Betriebswirt (FH)
Geschäftsführer

Tel:    +49 3641 384 910 0
Mobil:  +49 175 701 2201
***@***.***
Web:www.datainmotion.de  

Data In Motion Consulting GmbH
Kahlaische Straße 4
07745 Jena

Geschäftsführer
Mark Hoffmann
Jürgen Albert

Jena HRB 513025
Steuernummer 162/107/05779
USt-Id DE310002614

**Comment by laeubi** [comment]
> My POV would be different.

So can you then explain when one would use `@RequireServiceComponentRuntime` but _NOT_ `@Component`? Seems quite obsolete to me then...

> If I find code with component annotations, I would expect to have SCR in
the runtime.

Optional will not mean that SCR will not be available, it will just be included if possible to include one, but it is not mandatory to _resolve_ the bundle. So the main point is, why should my bundle _not resolve_ because SCR is not there? there is no reason for this.... Because in `resolved` state there won't be any components anyways.

> If I just find a Java class, but the Manifest declaration for components and the XML, I would probably think about why no annotations are used?

For me annotations and xml do not matter much its both the same in the end, but one mostly do not want to write the xml by hand. But then with annotations I'm forced to require scr all the time or use PDE because it currently do not add such header when generating XMLs.

**Comment by timothyjward** [comment]
> Why is is incorrect?
For the reason outlined where you cut my message off. Despite the theoretical issue you describe most bundles using SCR do not use it optionally. 

> Why is it hard to use? Actually nothing changes here in its usage...

To repeat In the overwhelming majority of cases (and every case I have ever used or implemented) a class declared `@Component` is not optional and a missing SCR runtime means that the bundle will not work. We spent years trying to fix this, and the current state is by far the best solution.

Unpicking this as you suggest to make the requirement optional will:
1. Break existing usage that relies on it being mandatory
2. Be *incorrect* metadata for most users where it should be a mandatory requirement 
3. Allow the framework to not wire the requirement to an extender, breaking the “multiple extender” tie break in the DS spec

> it does not mandates that a component must declare the requirement.
> …
> So the spec says it should do so and it can use but it never states it must be used.

Correct. The use of annotations is not required, you could add the requirement in other ways if you want. 
It is also not mandatory to include the extender requirement as the DS spec predates generic requirements and capabilities. Making it mandatory would have been a breaking change for existing bundles. It is therefore a *strong recommendation* but not a strict requirement. 

> If I have a BundleActivator, I mostly want my bundle to be started, still it is not required and I can still use classes from that bundle as it resolves

This is a good argument in favour of the status quo. If you do not start a bundle containing DS conmponents then the components do not start (just like an activator). In order to resolve the bundle’s requirements must be satisfied - in the case of the activator this includes the potentially unused package imports from the BundleAcivator implementation, much like the potentially unused SCR extender requirement.

**Comment by timothyjward** [comment]
TLDR - the theoretical bundles are too big and the real problem is that they mix API/library and implementation in the same module. Either accept that by making a monolith you need all your dependencies all of the time, or stop building monoliths.

**Comment by laeubi** [comment]
> Allow the framework to not wire the requirement to an extender, breaking the “multiple extender” tie break in the DS spec

If there is no SCR how could there be multiple extenders?

> This is a good argument in favour of the status quo. If you do not start a bundle containing DS conmponents then the components do not start (just like an activator). In order to resolve the bundle’s requirements must be satisfied - in the case of the activator this includes the potentially unused package imports from the BundleAcivator implementation, much like the potentially unused SCR extender requirement.

Package requirements are not mandatory for an activator, they are mandatory for resolving and creating a classloader and as any other bundle can load any class there is literally no "unused import", so I think the scr case is not really comparable as is not required here. So unless my bundle gets (lazy) started the requirement do not applies here, and even if it won't fail (just probably miss some of its functionality)

> top building monoliths

We have discussed that in the last spec call already, of course I can make `one class = one bundle` to be extra "modular", but in reality there is always a trade-off and there are always times where code needs to be transition or evolve.
And then there are practical projects, that don't care much about OSGi but are very concerned because things are "suddenly" required and break former uses and then better drop OSGi as a whole or stay at ancient versions/techniques.

**Comment by maho7791** [comment]
Hi Christoph,

Am 10.02.23 um 09:20 schrieb Christoph Läubrich:
>
>     My POV would be different.
>
> So can you then explain when one would use 
> ***@***.***| but /NOT/ ***@***.***|? Seems quite 
> obsolete to me then...
>
If you have bundle JARs in my repository that just have components and 
my code does not declare any components using annotations. Then I would 
use ***@***.***|, so that my application can 
resolve SCR|.||
|
>
>     If I find code with component annotations, I would expect to have
>     SCR in
>     the runtime.
>
> Optional will not mean that SCR will not be available, it will just be 
> included if possible to include one, but it is not mandatory to 
> /resolve/ the bundle. So the main point is, why should my bundle /not 
> resolve/ because SCR is not there? there is no reason for this.... 
> Because in |resolved| state there won't be any components anyways.
>
I try to describe my expectation in a different way. When I want to 
create a component using annotations, I want them to work with SCR. This 
means I want the resolver to also make SCR available, resolving the 
requirement for the extender.

If I want not to have SCR in place for whatever reasons, I would 
probably rather package the component XML by hand and taking car not 
having the SCR extender requirement in my Manifest.

And as described before, if I just orchestrate external JAR's in my 
application, I would use the ***@***.***| annotion, 
to ensure that SCR will be resolved.

>     If I just find a Java class, but the Manifest declaration for
>     components and the XML, I would probably think about why no
>     annotations are used?
>
> For me annotations and xml do not matter much its both the same in the 
> end, but one mostly do not want to write the xml by hand. But then 
> with annotations I'm forced to require scr all the time or use PDE 
> because it currently do not add such header when generating XMLs.
>
The hardest part in writing spec in getting many different views and 
POV's in one paper. And we have just two different views ;-)

Mark


-- 
Mark Hoffmann
M.A. Dipl.-Betriebswirt (FH)
Geschäftsführer

Tel:    +49 3641 384 910 0
Mobil:  +49 175 701 2201
***@***.***
Web:www.datainmotion.de  

Data In Motion Consulting GmbH
Kahlaische Straße 4
07745 Jena

Geschäftsführer
Mark Hoffmann
Jürgen Albert

Jena HRB 513025
Steuernummer 162/107/05779
USt-Id DE310002614

**Comment by timothyjward** [comment]
> If there is no SCR how could there be multiple extenders?

Because the framework in this theoretical use case is not the only framework in existence and the specification needs to work for as many people as possible. The resulting specification may not be perfectly optimal for a specific use case in order to support other valid use cases, or to avoid undefined behaviour. Specifications may also take a position to ensure ease of correct use in common cases at the expense of ease of use in uncommon cases.

> Package requirements are not mandatory for an activator, they are mandatory for resolving and creating a classloader

In order to load your activator class there must be an import for `org.osgi.framework` even if that package isn’t use anywhere else in the bundle. This is exactly analogous to a requirement for a DS extender.

**Comment by timothyjward** [comment]
In case it helps, bnd will let you do what you want right now with some extra configuration. 

The [-bundleannotations](https://bnd.bndtools.org/instructions/bundleannotations.html) instruction can be told to ignore one or more classes:

    -bundleannotations: !some.package.MyClass, \
      *

from there you can add in whatever requirements you need somewhere else (perhaps the package-info.java)

I see this as a far preferable solution to breaking most use cases to support this edge case.

**Comment by tjwatson** [comment]
> If I have a BundleActivator, I mostly want my bundle to be started, still it is not required and I can still use classes from that bundle as it resolves

If I have a `BundleActivator`, then I must import the `org.osgi.framework` package.  In order to resolve there must be an implementation of that package (the framework itself).  The contract of importing the package is that a framework implementation will call the specified bundle activator when the framework is activated.  There is no optional behavior in that respect.

Similarly, a bundle with service components expects their service components to be acted upon when their bundle is started.  This is exactly the same non-optional requirement that a bundle importing `org.osgi.framework` would expect for its specified `Bundle-Activator`.

**Comment by laeubi** [comment]
Alright, I agree that this is maybe not useful then on the long term.
